File: platform/android/lynx_android/src/main/java/com/lynx/tasm/gesture/handler/LongPressGestureHandler.java
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.lynx.tasm.gesture.handler;

import android.os.Handler;
import android.os.Looper;
import android.view.MotionEvent;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.lynx.react.bridge.ReadableMap;
import com.lynx.tasm.behavior.LynxContext;
import com.lynx.tasm.event.LynxTouchEvent;
import com.lynx.tasm.gesture.GestureArenaMember;
import com.lynx.tasm.gesture.detector.GestureDetector;
import com.lynx.tasm.utils.PixelUtils;

public class LongPressGestureHandler extends BaseGestureHandler {
  // Minimum time, expressed in milliseconds, that a finger must remain pressed on the corresponding
  // view. The default value is 500.
  private long mMinDuration = 500L;
  // Maximum distance in X and Y axis,  If the finger travels further than the defined distance
  // along the X axis and the gesture hasn't yet activated, it will fail to recognize the gesture.
  private float mMaxDistance = PixelUtils.dipToPx(10);

  private Handler mHandler = null;

  // record x axis when action down
  private float mStartX = 0f;
  // record y axis when action down
  private float mStartY = 0f;
  // record x axis when action up
  private float mLastX = 0f;
  // record x axis when action up
  private float mLastY = 0f;
  // record lynx touch event
  private LynxTouchEvent mLynxTouchEvent = null;
  // is onEnd invoked or not
  private boolean mIsInvokedEnd = false;

  private final Runnable mDelayActivateRunnable = () -> {
    if (mStatus != GestureConstants.LYNX_STATE_FAIL
        && mStatus != GestureConstants.LYNX_STATE_ACTIVE) {
      activate();
      onStart(mLastX, mLastY, mLynxTouchEvent);
    }
  };

  public LongPressGestureHandler(int sign, LynxContext lynxContext,
      @NonNull GestureDetector gestureDetector, @NonNull GestureArenaMember gestureArenaMember) {
    super(sign, lynxContext, gestureDetector, gestureArenaMember);
    handleConfigMap(gestureDetector.getConfigMap());
  }

  @Override
  protected void handleConfigMap(@Nullable ReadableMap config) {
    if (config == null) {
      return;
    }
    mMinDuration = config.getLong(GestureConstants.MIN_DURATION, 500);
    mMaxDistance = PixelUtils.dipToPx(config.getLong(GestureConstants.MAX_DISTANCE, 10));
  }

  private void startLongPress() {
    if (mHandler == null) {
      mHandler = new Handler(Looper.getMainLooper());
    } else {
      mHandler.removeCallbacksAndMessages(null);
    }
    mHandler.postDelayed(mDelayActivateRunnable, mMinDuration);
  }

  private void endLongPress() {
    if (mHandler != null) {
      mHandler.removeCallbacksAndMessages(null);
      mHandler = null;
    }
  }

  @Override
  protected void onHandle(@Nullable MotionEvent event, @Nullable LynxTouchEvent lynxTouchEvent,
      float flingDeltaX, float flingDeltaY) {
    mLynxTouchEvent = lynxTouchEvent;
    // If the event is empty, it means the finger not touches the screen
    if (event == null) {
      ignore();
      return;
    }

    if (mStatus >= GestureConstants.LYNX_STATE_FAIL) {
      endLongPress();
      return;
    }
    switch (event.getActionMasked()) {
      case MotionEvent.ACTION_DOWN:
        mStartX = event.getX();
        mStartY = event.getY();
        mIsInvokedEnd = false;
        begin();
        onBegin(mStartX, mStartY, lynxTouchEvent);
        startLongPress();
        break;
      case MotionEvent.ACTION_MOVE:
        mLastX = event.getX();
        mLastY = event.getY();
        if (shouldFail()) {
          fail();
          endLongPress();
        }
        break;
      case MotionEvent.ACTION_UP:
        endLongPress();
        fail();
        break;
      default:
        break;
    }
  }

  private boolean shouldFail() {
    float dx = Math.abs(mLastX - mStartX);
    float dy = Math.abs(mLastY - mStartY);
    if (dx > mMaxDistance || dy > mMaxDistance) {
      return true;
    }
    return false;
  }

  @Override
  public void fail() {
    if (mStatus != GestureConstants.LYNX_STATE_FAIL) {
      mStatus = GestureConstants.LYNX_STATE_FAIL;
      onEnd(mLastX, mLastY, mLynxTouchEvent);
    }
  }

  @Override
  public void end() {
    if (mStatus != GestureConstants.LYNX_STATE_END) {
      mStatus = GestureConstants.LYNX_STATE_END;
      onEnd(mLastX, mLastY, mLynxTouchEvent);
    }
  }

  @Override
  public void reset() {
    super.reset();
    mIsInvokedEnd = false;
  }

  @Override
  protected void onBegin(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnBeginEnable()) {
      return;
    }
    sendGestureEvent(GestureConstants.ON_BEGIN, getEventParamsFromTouchEvent(event));
  }

  @Override
  protected void onUpdate(float deltaX, float deltaY, @Nullable LynxTouchEvent event) {
    // empty implementation, because long press gesture is not continuous gesture
  }

  @Override
  protected void onStart(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnStartEnable()) {
      return;
    }
    sendGestureEvent(GestureConstants.ON_START, getEventParamsFromTouchEvent(event));
  }

  @Override
  protected void onEnd(float x, float y, @Nullable LynxTouchEvent event) {
    if (!isOnEndEnable() || mIsInvokedEnd) {
      return;
    }
    mIsInvokedEnd = true;
    sendGestureEvent(GestureConstants.ON_END, getEventParamsFromTouchEvent(event));
  }
}

File: platform/harmony/lynx_harmony/src/main/cpp/gesture/handler/longpress_gesture_handler.cc
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "platform/harmony/lynx_harmony/src/main/cpp/gesture/handler/longpress_gesture_handler.h"

#include "base/include/thread/timed_task.h"
#include "platform/harmony/lynx_harmony/src/main/cpp/event/gesture_event.h"
#include "platform/harmony/lynx_harmony/src/main/cpp/event/touch_event.h"
#include "platform/harmony/lynx_harmony/src/main/cpp/gesture/gesture_arena_member.h"
#include "platform/harmony/lynx_harmony/src/main/cpp/lynx_context.h"

namespace lynx {
namespace tasm {
namespace harmony {
LongPressGestureHandler::LongPressGestureHandler(
    int sign, LynxContext* lynx_context,
    std::shared_ptr<GestureDetector> gesture_detector,
    std::weak_ptr<GestureArenaMember> gesture_arena_member)
    : BaseGestureHandler(sign, lynx_context, gesture_detector,
                         gesture_arena_member),
      max_distance_(10),
      min_duration_(500),
      start_x_(0),
      start_y_(0),
      last_x_(0),
      last_y_(0),
      is_invoked_end_(false),
      timer_task_manager_(std::make_unique<base::TimedTaskManager>()) {
  HandleConfigMap(gesture_detector->gesture_config());
}

void LongPressGestureHandler::HandleConfigMap(const lepus::Value& config) {
  if (config.Contains(GestureConstants::MIN_DURATION)) {
    min_duration_ = config.GetProperty(GestureConstants::MIN_DURATION).Number();
  }
  if (config.Contains(GestureConstants::MAX_DISTANCE)) {
    max_distance_ =
        Dip2Px(config.GetProperty(GestureConstants::MAX_DISTANCE).Number());
  }
}

void LongPressGestureHandler::OnHandle(
    const ArkUI_UIInputEvent* event,
    std::shared_ptr<TouchEvent> lynx_touch_event, float fling_delta_x,
    float fling_delta_y) {
  last_touch_event_ = lynx_touch_event;
  if (event == nullptr) {
    Ignore();
    return;
  }
  if (status_ >= GestureConstants::LYNX_STATE_FAIL) {
    EndLongPress();
    return;
  }

  switch (OH_ArkUI_UIInputEvent_GetAction(event)) {
    case UI_TOUCH_EVENT_ACTION_DOWN:
      start_x_ = OH_ArkUI_PointerEvent_GetXByIndex(event, 0);
      start_y_ = OH_ArkUI_PointerEvent_GetYByIndex(event, 0);
      is_invoked_end_ = false;
      Begin();
      OnBegin(start_x_, start_y_, lynx_touch_event);
      StartLongPress();
      break;
    case UI_TOUCH_EVENT_ACTION_MOVE:
      last_x_ = OH_ArkUI_PointerEvent_GetXByIndex(event, 0);
      last_y_ = OH_ArkUI_PointerEvent_GetYByIndex(event, 0);
      if (ShouldFail()) {
        Fail();
        EndLongPress();
      }
      break;
    case UI_TOUCH_EVENT_ACTION_UP:
      EndLongPress();
      Fail();
      break;
    default:
      break;
  }
}

void LongPressGestureHandler::Fail() {
  if (status_ != GestureConstants::LYNX_STATE_FAIL) {
    status_ = GestureConstants::LYNX_STATE_FAIL;
    OnEnd(last_x_, last_y_, last_touch_event_);
  }
}

void LongPressGestureHandler::End() {
  if (status_ != GestureConstants::LYNX_STATE_END) {
    status_ = GestureConstants::LYNX_STATE_END;
    OnEnd(last_x_, last_y_, last_touch_event_);
  }
}

void LongPressGestureHandler::Reset() {
  BaseGestureHandler::Reset();
  is_invoked_end_ = false;
}
void LongPressGestureHandler::StartLongPress() {
  auto task = [this]() {
    if (status_ != GestureConstants::LYNX_STATE_FAIL &&
        status_ != GestureConstants::LYNX_STATE_ACTIVE) {
      Activate();
      OnStart(last_x_, last_y_, last_touch_event_);
    }
  };
  last_task_id_ =
      timer_task_manager_->SetTimeout(std::move(task), min_duration_);
}

void LongPressGestureHandler::EndLongPress() {
  timer_task_manager_->StopTask(last_task_id_);
}

bool LongPressGestureHandler::ShouldFail() {
  float dx = std::abs(last_x_ - start_x_);
  float dy = std::abs(last_y_ - start_y_);
  return dx > max_distance_ || dy > max_distance_;
}

void LongPressGestureHandler::OnBegin(float x, float y,
                                      std::shared_ptr<TouchEvent> event) {
  if (!IsOnBeginEnable()) {
    return;
  }
  SendGestureEvent(GestureConstants::ON_BEGIN,
                   GetEventParamsFromTouchEvent(event));
}

void LongPressGestureHandler::OnUpdate(float delta_x, float delta_y,
                                       std::shared_ptr<TouchEvent> event) {
  // empty implementation, because long press gesture is not continuous gesture
}

void LongPressGestureHandler::OnStart(float x, float y,
                                      std::shared_ptr<TouchEvent> event) {
  if (!IsOnStartEnable()) {
    return;
  }
  SendGestureEvent(GestureConstants::ON_START,
                   GetEventParamsFromTouchEvent(event));
}

void LongPressGestureHandler::OnEnd(float x, float y,
                                    std::shared_ptr<TouchEvent> event) {
  if (!IsOnEndEnable() || is_invoked_end_) {
    return;
  }
  is_invoked_end_ = true;
  SendGestureEvent(GestureConstants::ON_END,
                   GetEventParamsFromTouchEvent(event));
}

}  // namespace harmony
}  // namespace tasm
}  // namespace lynx
